### ISSUE TITLE
For anyone who needs a Python 3-compatible FormAlchemy that actually works, read here...

### ISSUE THREAD

**Issue by coredumperror** [FormAlchemy/formalchemy#71]
So if you're like me, and needed to update an ancient webapp that uses FormAlchemy, but were frustrated by the lack of a release that fixes the WebHelpers incompatibility which causes an error related to not being able to use lists/tuples for options, here's how to fix that:

```
git clone git@github.com:FormAlchemy/formalchemy.git
cd formalchemy
python setup.py sdist
cd dist
```

You're now in a folder with a file called `FormAlchemy-1.6.dev0.tar.gz`. Put that file somewhere online (a S3 bucket works great), and then add that URL to your requirements file. You _may_ also need to update from `WebHelpers==1.3` to `WebHelpers2==2.0`, but I'm not 100% certain about that.

**Comment by aconrad** [comment]
Hi Robert,

Thanks for sharing. Are you saying that master has an unreleased version
that has the fix you need? And if so then all you need would be a new
release of FormAlchemy?

I'm not sure who's in charge of releases now, but since you felt the pain
then if someone reading this has the ability to do releases then it sounds
like it would make a positive impact to some folks out there. Anyone?

**Comment by coredumperror** [comment]
Yes, master has a number of changes that did not make it into 1.5.5, including one that makes it compatible with a change in WebHelpers from several years ago, which completely broke FormAlchemy.

Also, something I forgot to mention is that line 7 of setup.py needs to be removed, because the six library is not available during the execution of setup.py while installing FormAlchemy through pip. It does an import that doesn't even get used anyway, so there's no harm in removing it.

**Comment by sallner** [comment]
@coredumperror Maybe you could add a PR to make that change easier to include in the next release?

@gawel @aconrad @smurfix Do you see a chance to cut a release?

**Comment by coredumperror** [comment]
Sure, I could do that. To be honest, I expected my comment to sit here, unread, for possibly years, until someone like myself came along and tried to update an ancient codebase that used FormAlchemy. If I'd know that not one, but two people would actually come along and see it within a day, I would have made a PR.

So I'll go do that. Though I believe it'll just be that one change to setup.py, as there are no other changes necessary on the master branch to make a working version.

**Comment by aconrad** [comment]
Thank you @coredumperror, this is much appreciated. Once it gets merged, I'll see if I have enough permissions to issue a new release.

**Comment by sallner** [comment]
@coredumperror What do you think, will you need to support your app for a longer period of time? I.e., is it reasonable to keep FormAlchemy in a state that is reliably tested via CI? I did not managed to run a buildout with successful tests afterwards, so there might be some work involved beyond the usual cleanup.

**Comment by sallner** [comment]
@aconrad Would you cut a release, as I do not have any PyPI rights for this package. There is a guide to create a release in the repo, although I failed with the tests.

**Comment by aconrad** [comment]
@sallner thanks for merging the PR. Could you do the release if I gave you the necessary PyPI permissions to publish a new FA release?

**Comment by sallner** [comment]
@aconrad Yes, I would do that.

**Comment by aconrad** [comment]
What is your pypi username?

**Comment by sallner** [comment]
it is `sallner`, just like in github.

**Comment by sallner** [comment]
@aconrad Have you tried to give me permission on PyPI?

**Comment by smurfix** [comment]
On 10.11.20 09:04, Steffen Allner wrote:
>
> @aconrad <https://github.com/aconrad> Have you tried to give me 
> permission on PyPI?
>
I just added you, you should have received an invite.

-- 
-- Matthias Urlichs

**Comment by sallner** [comment]
Thank you, that worked. I will prepare a release.

**Comment by aconrad** [comment]
Sorry for the late reply, and thanks Matthias for that.

**Comment by coredumperror** [comment]
Glad to hear that a working official release is coming!

**Comment by sallner** [comment]
@coredumperror I made a release 1.5.6 https://pypi.org/project/FormAlchemy/1.5.6/. But I have to say, that the 1.5.5 already included the fix of the wrong `six` import in `setup.py`. The respective code was found without tag on the `1.5` branch. Please give it a try, whether it still works. 

Sidenote: Do you plan to use FormAlchemy for a longer time? Do we need to activate a CI? It is rather difficult to clean the repo and make tests working again.

**Comment by coredumperror** [comment]
> the 1.5.5 already included the fix of the wrong six import in setup.py.

Yes, but it didn't include the *other* changes made to master since 1.5.5 was released. Those were already working in builds made manually from master. I only did the PR for this fix because it made it possible to make a new, working PyPI release that has those changes.

I don't believe there's any good reason to make any further FormAlchemy releases. At least not for my sake. The only reason I came to this issue queue was because I'm working on updating an ancient app that my team hasn't got time to rewrite from scratch using modern tech. We wanted to update the app to Python 3, which meant we absolutely needed to update our version of FormAlchemy from 1.4.3. But I found that the latest version on PyPI was broken due to a later change made to one of its dependencies. Now that that's fixed, FormAlchemy should be pretty stable.

**Comment by aconrad** [comment]
Thanks @sallner for making the release, much appreciated! @coredumperror thanks again for your report and everyone else who chimed in. We're making people's work a little better this way.

**Comment by coredumperror** [comment]
There's still a huge problem, as FormAlchemy 1.5.6 cannot be installed on Python 3, due to an import error in MsgFmt.py. cStringIO doesn't exist in Python 3.

I'm pretty sure the issue is due to the changes in `setup.py` related to compiling translated messages during setup. I have no idea how to deal with that, though, besides ripping it out entirely (which I tried, and it works). 

I tried editing MsgFmt.py to make it Python3 compatible, but I the more editing I did, the more it broke. I eventually gave up. I have no idea what MsgFmt.py is even doing, or why. Maybe there's a better way to do what it's doing in modern Python? It seems *very* antiquated, from the glimpses I took at the code while fixing the string of new exceptions that kept popping up.

**Comment by smurfix** [comment]
It's an ancient replacement for "msgfmt". Just call the system's msgfmt program.

**Comment by sallner** [comment]
So we need tests anyhow. It is never a good idea to run without it and we should enforce it here again.

**Comment by smurfix** [comment]
I pushed a rudimentary "call the system's msgfmt" PR. Could be better but does the job for now.

**Comment by shasan-beanfield** [comment]
any updates on this?  It's stating to affect us upgrading to latest version of python
